File: scripts/images/crop_menu_items.py
import cv2
import os
from PIL import Image

# ── PATH CONFIGURATION ──
# Use relative paths from script location
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
MENU_IMAGE = os.path.join(SCRIPT_DIR, '..', '..', 'public', 'assets', 'menu_full.jpg')
OUTPUT_DIR = os.path.join(SCRIPT_DIR, '..', '..', 'public', 'assets')

# Ensure output directory exists
os.makedirs(OUTPUT_DIR, exist_ok=True)

# ── MENU ITEMS & THEIR CROP COORDINATES (x, y, width, height) ──
# These coordinates are approximate and should be adjusted based on your actual image
CROP_ITEMS = {
    # CHICKEN MENU
    'chicken_burger.jpg': (375, 100, 170, 170),  # Chicken Burger
    'chicken_burger_chips.jpg': (560, 250, 170, 170),  # Chicken Burger & Chips
    'double_chicken_chips.jpg': (375, 250, 170, 170),  # Double Chicken Burger & Chips
    'chicken_fries_fanta.jpg': (375, 250, 170, 170),  # Chicken Fries & Fanta
    'omelette_burger_chips.jpg': (560, 100, 170, 170),  # Omelette Burger & Chips
    
    # BOX CHICKEN COMBOS
    'chicken_combo_1.jpg': (375, 250, 170, 170),  # Chicken Burger & Chips & Fanta
    'chicken_combo_2.jpg': (560, 250, 170, 170),  # Double Chicken Burger & Chips & Fanta
    'chicken_fries_combo.jpg': (375, 250, 170, 170),  # Chicken Fries & Fanta
    
    # BEEF MENU
    'beef_burger.jpg': (375, 350, 170, 170),  # Beef Burger
    'beef_burger_chips.jpg': (560, 350, 170, 170),  # Beef Burger & Chips
    'double_beef_chips.jpg': (375, 500, 170, 170),  # Double Beef Burger & Chips
    'beef_fries.jpg': (560, 350, 170, 170),  # Beef Fries
    
    # BOX BEEF COMBOS
    'beef_combo_1.jpg': (375, 450, 170, 170),  # Beef Burger & Chips & Fanta
    'beef_combo_2.jpg': (560, 450, 170, 170),  # Double Beef Burger & Chips & Fanta
    'beef_fries_combo.jpg': (375, 550, 170, 170),  # Beef Fries & Fanta
    
    # EXTRAS
    'sauces_cheese.jpg': (375, 600, 170, 170),  # Sauces & Cheese
    'fanta.jpg': (560, 600, 170, 170),  # Fanta
    'chips.jpg': (375, 650, 170, 170),  # Chips
}

def crop_menu_items():
    """Crop all menu items from the menu image"""
    
    if not os.path.exists(MENU_IMAGE):
        print(f"❌ Menu image not found at: {MENU_IMAGE}")
        print("📝 Please save your menu image as 'menu_full.jpg' in the assets folder")
        return
    
    # Load the menu image
    img = cv2.imread(MENU_IMAGE)
    if img is None:
        print(f"❌ Could not read image: {MENU_IMAGE}")
        return
    
    height, width = img.shape[:2]
    print(f"📷 Image size: {width}x{height}")
    print(f"🔄 Starting to crop {len(CROP_ITEMS)} menu items...")
    print()
    
    success_count = 0
    
    for filename, (x, y, w, h) in CROP_ITEMS.items():
        # Crop the region
        cropped = img[y:y+h, x:x+w]
        
        # Save the cropped image
        output_path = os.path.join(OUTPUT_DIR, filename)
        cv2.imwrite(output_path, cropped)
        
        success_count += 1
        print(f"✅ {success_count}. {filename}")
    
    print()
    print(f"🎉 Successfully cropped {success_count} images!")
    print(f"📁 Saved to: {OUTPUT_DIR}")

def calibrate_crops():
    """Interactive tool to find exact coordinates"""
    
    if not os.path.exists(MENU_IMAGE):
        print(f"❌ Menu image not found at: {MENU_IMAGE}")
        return
    
    img = cv2.imread(MENU_IMAGE)
    if img is None:
        print(f"❌ Could not read image: {MENU_IMAGE}")
        return
    
    print("🎯 CROP CALIBRATION MODE")
    print("=" * 50)
    print("Click and drag to select a region, then press 'c' to copy coordinates")
    print("Press 'q' to quit calibration")
    print()
    
    # Display image with mouse callback
    cv2.imshow('Menu Items - Select regions', img)
    cv2.setMouseCallback('Menu Items - Select regions', mouse_callback)
    
    while True:
        key = cv2.waitKey(1) & 0xFF
        if key == ord('q'):
            break
    
    cv2.destroyAllWindows()

def mouse_callback(event, x, y, flags, param):
    """Mouse callback for interactive calibration"""
    if event == cv2.EVENT_LBUTTONDOWN:
        print(f"Start: ({x}, {y})")
    elif event == cv2.EVENT_RBUTTONDOWN:
        print(f"End: ({x}, {y})")

if __name__ == '__main__':
    print("=" * 60)
    print("🔥 MAD BURNING - MENU ITEM CROPPER 🔥")
    print("=" * 60)
    print()
    
    # Run the cropper
    crop_menu_items()
    
    print()
    print("📌 MANUAL ADJUSTMENT NEEDED:")
    print("The coordinates above are approximate.")
    print("You may need to adjust them in the CROP_ITEMS dictionary")
    print("based on your actual image dimensions.")
    print()
    print("💡 TIP: Use the calibrate_crops() function to find exact coordinates")

File: scripts/images/crop_menu_items.js
#!/usr/bin/env node

/**
 * 🔥 MAD BURNING - MENU ITEM CROPPER (Node.js Version) 🔥
 * Automatically crops product images from menu grid
 */

const fs = require('fs');
const path = require('path');

// Install required package first
function checkDependencies() {
  try {
    require('sharp');
  } catch (e) {
    console.log('📦 Installing required dependency: sharp');
    console.log('Running: npm install sharp\n');
    require('child_process').execSync('npm install sharp', { stdio: 'inherit' });
  }
}

async function cropMenuItems() {
  const sharp = require('sharp');
  
  const MENU_IMAGE = path.join(__dirname, '..', '..', 'public', 'assets', 'menu_full.png');
  const OUTPUT_DIR = path.join(__dirname, '..', '..', 'public', 'assets');
  
  // Check if image exists
  if (!fs.existsSync(MENU_IMAGE)) {
    console.error('❌ ERROR: menu_full.png not found');
    console.error(`Expected path: ${MENU_IMAGE}\n`);
    console.log('📝 STEPS:');
    console.log('1. Save your menu image as menu_full.png');
    console.log('2. Place it in: public/assets/\n');
    process.exit(1);
  }
  
  // Menu items with estimated crop coordinates (x, y, width, height)
  // These are approximate based on typical 5-column grid layout
  const CROP_ITEMS = [
    // Row 1: Plain items (no Fanta)
    { file: 'chicken_burger.jpg', name: 'Chicken Burger', x: 20, y: 40, w: 140, h: 140 },
    { file: 'beef_burger.jpg', name: 'Beef Burger', x: 170, y: 40, w: 140, h: 140 },
    { file: 'double_chicken_burger.jpg', name: 'Double Chicken Burger', x: 320, y: 40, w: 140, h: 140 },
    { file: 'fries.jpg', name: 'Fries', x: 470, y: 40, w: 140, h: 140 },
    { file: 'omelette_burger.jpg', name: 'Omelette Burger', x: 620, y: 40, w: 140, h: 140 },
    
    // Row 2: With Fanta
    { file: 'chicken_burger_fanta.jpg', name: 'Chicken Burger & Fanta', x: 20, y: 200, w: 140, h: 140 },
    { file: 'beef_burger_fanta.jpg', name: 'Beef Burger & Fanta', x: 170, y: 200, w: 140, h: 140 },
    { file: 'double_chicken_fanta.jpg', name: 'Double Chicken & Fanta', x: 320, y: 200, w: 140, h: 140 },
    { file: 'fries_fanta.jpg', name: 'Fries & Fanta', x: 470, y: 200, w: 140, h: 140 },
    { file: 'cheese_burger_fanta.jpg', name: 'Cheese Burger & Fanta', x: 620, y: 200, w: 140, h: 140 },
    
    // Row 3: Beef variants
    { file: 'beef_cheese_burger.jpg', name: 'Beef Cheese Burger', x: 20, y: 360, w: 140, h: 140 },
    { file: 'beef_burger_cheese.jpg', name: 'Beef Burger Cheese', x: 170, y: 360, w: 140, h: 140 },
    { file: 'fries_plain.jpg', name: 'Plain Fries', x: 320, y: 360, w: 140, h: 140 },
    { file: 'double_beef_burger.jpg', name: 'Double Beef Burger', x: 470, y: 360, w: 140, h: 140 },
    { file: 'loaded_fries.jpg', name: 'Loaded Fries', x: 620, y: 360, w: 140, h: 140 },
    
    // Row 4: More combos
    { file: 'beef_combo_1.jpg', name: 'Beef Combo 1', x: 20, y: 520, w: 140, h: 140 },
    { file: 'loaded_fries_2.jpg', name: 'Loaded Fries 2', x: 170, y: 520, w: 140, h: 140 },
    { file: 'beef_fanta_combo.jpg', name: 'Beef & Fanta Combo', x: 320, y: 520, w: 140, h: 140 },
    { file: 'double_beef_fanta.jpg', name: 'Double Beef & Fanta', x: 470, y: 520, w: 140, h: 140 },
    { file: 'loaded_fries_fanta.jpg', name: 'Loaded Fries & Fanta', x: 620, y: 520, w: 140, h: 140 },
    
    // Row 5: Extras
    { file: 'sauces.jpg', name: 'Sauces & Condiments', x: 20, y: 680, w: 140, h: 140 },
    { file: 'cheese_mayo.jpg', name: 'Cheese & Mayo', x: 170, y: 680, w: 140, h: 140 },
    { file: 'fanta_drink.jpg', name: 'Fanta Drink', x: 320, y: 680, w: 140, h: 140 },
    { file: 'fries_strips.jpg', name: 'Fries Strips', x: 470, y: 680, w: 140, h: 140 },
    { file: 'fries_container.jpg', name: 'Fries Container', x: 620, y: 680, w: 140, h: 140 },
  ];
  
  console.log('\n' + '='.repeat(70));
  console.log('🔥 MAD BURNING - MENU ITEM CROPPER 🔥');
  console.log('='.repeat(70) + '\n');
  
  console.log(`📷 Processing: ${MENU_IMAGE}`);
  console.log(`📁 Output to: ${OUTPUT_DIR}\n`);
  
  // Get image metadata
  const metadata = await sharp(MENU_IMAGE).metadata();
  console.log(`✅ Image dimensions: ${metadata.width}x${metadata.height}px\n`);
  console.log(`🔄 Cropping ${CROP_ITEMS.length} items...\n`);
  console.log('='.repeat(70));
  
  let successCount = 0;
  let errors = [];
  
  // Crop each item
  for (const item of CROP_ITEMS) {
    try {
      const outputPath = path.join(OUTPUT_DIR, item.file);
      
      await sharp(MENU_IMAGE)
        .extract({
          left: item.x,
          top: item.y,
          width: item.w,
          height: item.h
        })
        .toFile(outputPath);
      
      successCount++;
      console.log(`✅ ${successCount.toString().padStart(2)}. ${item.name.padEnd(40)} → ${item.file}`);
    } catch (err) {
      errors.push({ item: item.name, error: err.message });
      console.log(`❌ ${item.name.padEnd(40)} → ERROR: ${err.message}`);
    }
  }
  
  console.log('='.repeat(70) + '\n');
  console.log(`🎉 Successfully cropped: ${successCount}/${CROP_ITEMS.length} images`);
  
  if (errors.length > 0) {
    console.log(`\n⚠️  Errors: ${errors.length}`);
    errors.forEach(e => {
      console.log(`   - ${e.item}: ${e.error}`);
    });
  }
  
  console.log(`\n📁 All images saved to: ${OUTPUT_DIR}\n`);
  
  if (successCount > 0) {
    console.log('✨ Next steps:');
    console.log('1. Refresh your website');
    console.log('2. All menu images will display automatically\n');
  }
}

// Main execution
async function main() {
  try {
    checkDependencies();
    await cropMenuItems();
  } catch (err) {
    console.error('❌ Error:', err.message);
    process.exit(1);
  }
}

main();
